amaranth_soc/csr/bus.py

```diff
@@ -126,50 +179,151 @@ class Interface(Record):
+        return self._data_width
+
+    @classmethod
+    def check(cls, *, addr_width, data_width):
+        """Validate signature parameters.
+
+        Raises
+        ------
+        :exc:`TypeError`
+            If ``addr_width`` is not an integer greater than 0.
+        :exc:`TypeError`
+            If ``data_width`` is not an integer greater than 0.
+        """
         if not isinstance(addr_width, int) or addr_width <= 0:
-            raise ValueError("Address width must be a positive integer, not {!r}"
-                             .format(addr_width))
+            raise TypeError(f"Address width must be a positive integer, not {addr_width!r}")
         if not isinstance(data_width, int) or data_width <= 0:
-            raise ValueError("Data width must be a positive integer, not {!r}"
-                             .format(data_width))
-        self.addr_width = addr_width
-        self.data_width = data_width
-        self._map       = None
-
-        super().__init__([
-            ("addr",    addr_width),
-            ("r_data",  data_width),
-            ("r_stb",   1),
-            ("w_data",  data_width),
-            ("w_stb",   1),
-        ], name=name, src_loc_at=1)
+            raise TypeError(f"Data width must be a positive integer, not {data_width!r}")
+
+    def create(self, *, path=()):
+        """Create a compatible interface.
+
+        See :meth:`wiring.Signature.create` for details.
+
+        Returns
+        -------
+        An :class:`Interface` object using this signature.
+        """
+        return Interface(self, path=path)
+
+    def __eq__(self, other):
+        """Compare signatures.
+
+        Two signatures are equal if they have the same address width and data width.
+        """
+        return (isinstance(other, Signature) and
+                self.addr_width == other.addr_width and
+                self.data_width == other.data_width)
+
+    def __repr__(self):
+        return f"csr.Signature({self.members!r})"
+
+
+class Interface(wiring.Interface):
+    """CPU-side CSR interface.
+
+    A low-level interface to a set of atomically readable and writable peripheral CSR registers.
+
+    Operation
+    ---------
+
+    CSR registers mapped to the CSR bus are split into chunks according to the bus data width.
+    Each chunk is assigned a consecutive address on the bus. This allows accessing CSRs of any
+    size using any datapath width.
+
+    When the first chunk of a register is read, the value of a register is captured, and reads
+    from subsequent chunks of the same register return the captured values. When any chunk except
+    the last chunk of a register is written, the written value is captured; a write to the last
+    chunk writes the captured value to the register. This allows atomically accessing CSRs larger
+    than datapath width.
+
+    Parameters
+    ----------
+    signature : :class:`Signature`
+        CPU-side CSR signature.
+    path : iter(:class:`str`)
```

Comment:
Likewise.

Reply:
Replaced with a cross-ref to `wiring.Interface`.